File: python/day14/knot.py
from typing import List


class Node:
    def __init__(
        self,
        value: int,
    ) -> None:
        self.value: int = value

    def set_next(self, next_: "Node") -> None:
        self.next: Node = next_

    def set_prev(self, prev: "Node") -> None:
        self.prev: Node = prev


class Elements:
    def __init__(self, size: int) -> None:
        self.size: int = size
        self.skip_size: int = 0
        self.current_postion: Node = Node(value=0)
        self.head: Node = self.current_postion
        prev = self.current_postion

        for value in range(1, self.size):
            n: Node = Node(value)
            n.set_prev(prev)
            prev.set_next(n)
            prev = n
        prev.set_next(self.current_postion)
        self.current_postion.set_prev(prev)

    def apply(self, length: int) -> None:

        current: Node = self.current_postion
        for _ in range(length - 1):
            current = current.next

        head: Node = self.current_postion
        tail: Node = current

        while head != tail and head.prev != tail:
            value: int = head.value

            head.value = tail.value
            tail.value = value

            head = head.next
            tail = tail.prev

        current = self.current_postion
        for _ in range(length + self.skip_size):
            current = current.next

        self.current_postion = current
        self.skip_size += 1


def get_lengths(data: str) -> List[int]:
    final_sequence: List[int] = [ord(c) for c in data]
    final_sequence.extend([17, 31, 73, 47, 23])

    return final_sequence


def hash(input_: str, size: int = 256) -> str:
    lengths: List[int] = get_lengths(input_)
    elements: Elements = Elements(size)

    for _ in range(64):
        for length in lengths:
            elements.apply(length)

    dense_hash: List[int] = []
    current: Node = elements.head
    for _ in range(16):
        xor: int = 0
        for _ in range(16):
            xor ^= current.value
            current = current.next
        dense_hash.append(xor)

    return "".join([f"{n:0{2}x}" for n in dense_hash])

File: go/day14_part2/knot.go
package main

import (
	"fmt"
	"strings"
)

type Elements struct {
	size             int
	skip_size        int
	current_position int
	elements         []int
	next             []int
	prev             []int
}

func mod(a, b int) int {
	return (a%b + b) % b
}

func MakeElements(size int) Elements {
	elements := []int{}
	next := []int{}
	prev := []int{}

	for i := 0; i < size; i++ {
		elements = append(elements, i)
		next = append(next, mod(i+1, size))
		prev = append(prev, mod(i-1, size))
	}
	return Elements{size, 0, 0, elements, next, prev}
}

func (e *Elements) apply(length int) {
	current_index := e.current_position
	for i := 0; i < length-1; i++ {
		current_index = e.next[current_index]
	}

	head_index := e.current_position
	tail_index := current_index

	for head_index != tail_index && e.prev[head_index] != tail_index {
		value := e.elements[head_index]

		e.elements[head_index] = e.elements[tail_index]
		e.elements[tail_index] = value

		head_index = e.next[head_index]
		tail_index = e.prev[tail_index]
	}

	current_index = e.current_position
	for i := 0; i < length+e.skip_size; i++ {
		current_index = e.next[current_index]
	}

	e.current_position = current_index
	e.skip_size += 1
}

func (e Elements) top() int {
	return e.elements[0] * e.elements[e.next[0]]
}

func get_lengths(data string) []int {
	sequence := []int{}
	for _, char := range data {
		sequence = append(sequence, int(char))
	}
	sequence = append(sequence, []int{17, 31, 73, 47, 23}...)
	return sequence
}

func knot_hash(input string, size int) string {
	lengths := get_lengths(input)
	e := MakeElements(size)

	for i := 0; i < 64; i++ {
		for _, length := range lengths {
			e.apply(length)
		}
	}
	dense_hash := []int{}
	current_index := 0
	for i := 0; i < 16; i++ {
		xor := 0
		for j := 0; j < 16; j++ {
			xor ^= e.elements[current_index]
			current_index = e.next[current_index]
		}
		dense_hash = append(dense_hash, xor)
	}

	output := []string{}
	for _, n := range dense_hash {
		output = append(output, fmt.Sprintf("%02x", n))
	}

	return strings.Join(output, "")
}
